MacOS native build fixes

piopawlu opened the pull request:
Fixes mouse and python support on native MacOS builds. I only changed the legacy renderer as the default/full one causes kernel panic after a few seconds of operation and I can't even debug it.

Milek7 commented:
Hmm, why does explicit link to system OpenGL is required? It shouldn't be required as OpenGL functions are loaded dynamically by glad loader.

macpython/macpython64 python locations were correct, these provided in python/python64 are Windows binaries. Now it likely just fails to load. (acquire_lock is indeed buggy it that it shouldn't crash when python is not loaded correctly, but that different issue). Python directories on all platforms should be generated through virtualenv, like that:
```
virtualenv2 macpython64
source macpython64/bin/activate
pip install pillow
```

Why it it necessary to guard GLFW_OPENGL_COMPAT_PROFILE setting with ifndef APPLE? What context you are getting without this? We specifically require 3.3 compatibility context because legacy renderer use opengl <2 calls, but at the same time we use opengl3 imgui implementation.

piopawlu commented:
I guess the python is my installation issue then, I don’t think the virtual environment is working on my machine. I will remove it from my PR and fix the virtual env instead.

Regarding OpenGL in cmake file I was getting linking errors without explicitly adding OpenGL. Can try dig a bit deeper next time I have a chance maybe it won’t be necessary.

Without the change for the legacy renderer I get the following error:
“NSGL: The targeted version of macOS only supports forward-compatible core profile contexts for OpenGL 3.2 and above" Could be because I am only targeting the latest release of MacOS.

I was getting a different error when I was trying the full rendered, but as it crashes my laptop even when it finally starts i will just revert that change.

Milek7 commented:
>Regarding OpenGL in cmake file I was getting linking errors without explicitly adding OpenGL. Can try dig a bit deeper next time I have a chance maybe it won’t be necessary.

That's curious, I think MacOS builds succeeded lately on Azure Pipelines without cmake changes. (now it's broken because it seems homebrew removed python2, ehh, I need to do something about it...)

> I was getting a different error when I was trying the full rendered, but as it crashes my laptop even when it finally starts i will just revert that change.

You can keep that change with GLFW_OPENGL_FORWARD_COMPAT, apparently MacOS requires it. I even think it should be safe to set on all platforms, not on APPLE only.
That's rather unfortunate that we have no way to debug why modern renderer crashes kernel. Did you remember to copy `shaders` directory from source tree to simulator workdir?

glfwGetWindowContentScale needs to be guarded by something like `#if GLFW_VERSION_MAJOR > 3 || (GLFW_VERSION_MAJOR == 3 && GLFW_VERSION_MINOR >= 3)` as it breaks build on older GLFW versions.

> “NSGL: The targeted version of macOS only supports forward-compatible core profile contexts for OpenGL 3.2 and above"

Hmm, that message means that legacy renderer shouldn't work at all, so this is bit weird that legacy works while it shouldn't, and modern crashes kernel but it should work.. :P

piopawlu commented:
I spent more time trying to investigate the problem and here's my update:

1. Commenting out glfwWindowHint for legacy renderer helped because it kept compatibility level at OpenGL 2.1. After enabling USE_IMGUI_GL3 and returning to OpenGL 3.x it builds, starts fine (GUI works ok) but nothing gets rendered. The same problem exists for both legacy and full renderers so it's not specific to any of them. From what I managed to find on Google it's usually a problem of not using OpenGL 3 the right way.. like mixing legacy and modern calls. I have no experience with OpenGL so I can't really comment on that. All shaders are in place,, and given that the code works on Windows I guess nobody cares about macOS enough to actually spend time on this. This could also be caused by shader incompatibility. I'd like to try the current production version code (20.09), but it's only in SVN and I don't have access there. Does it correspond to any commit in your branch?

2. The kernel panic still occurs every now and then,, I have a crash log but it's kernel stack only so don't think it will help here with anything. It's inside gfx card driver,, and Apple will of course not do anything with that.

3. When using USE_IMGUI_GL3 I no longer need to explicitly include OpenGL in CMake, but since the simulator is not usable on macOS when using OpenGL 3.x it doesn't really help.

Milek7 commented:
>Commenting out glfwWindowHint for legacy renderer helped because it kept compatibility level at OpenGL 2.1.

Fair enough, but it only makes sense for imgui gl2 implementation. I have relaxed it to request 2.0/3.0 context depending on built imgui version.

>I guess nobody cares about macOS enough to actually spend time on this

It's not really about time but I don't have access to Mac hardware, and it's not possible to test rendering on virtual machine. Could you test latest build from this repository, `sim` branch, with `USE_IMGUI_GL3`, copy `shaders` directory from repository into your maszyna files, with these .ini settings?
```
python.enabled no
gfxrenderer full
gfx.skippipeline yes
gfx.shadergamma yes
```

>I'd like to try the current production version code (20.09), but it's only in SVN and I don't have access there. Does it correspond to any commit in your branch?

What do you mean? MaSzyna 20.09 pack contained some older build from this repo, but it is not worth using old executables now. SVN contains newer game assets but they are not much relevant in debugging this.

>The kernel panic still occurs every now and then,, I have a crash log but it's kernel stack only so don't think it will help here with anything. It's inside gfx card driver,, and Apple will of course not do anything with that.

Kernel crashes could have security implications, so I guess Apple *should* care about them.

>When using USE_IMGUI_GL3 I no longer need to explicitly include OpenGL in CMake, but since the simulator is not usable on macOS when using OpenGL 3.x it doesn't really help

It might have been some extraneous includes in imgui2 builds, I think I fixed this.

I added that FORWARD_COMPAT flag change into repository.

Also I have fixed python disable flag, so `python.enabled no` in ini file should eliminate any python-related problems as it won't be initialized at all.

I also fixed Azure Pipelines builds and they should contain MacOS builds with any extra dependencies linked statically. While at it I have added ARM64 builds too (for Apple Silicon hardware).

Overall, everything this PR contained except mouse scaling should be already addressed by recent changes in this repository.

piopawlu committed:
Added fWindowXScale and fWindowYScale to support retina/hi-res displays on MacOS

piopawlu committed:
Fixed mouse handling for hi-res/retina displays on MacOS with legacy renderer

Milek7 commented:
CI fails.

We want to keep building with older GLFW possible, so glfwGetWindowContentScale needs to be gated like that:
```
#if GLFW_VERSION_MAJOR > 3 || (GLFW_VERSION_MAJOR == 3 && GLFW_VERSION_MINOR >= 3)
// glfw call there
#endif
```

piopawlu commented:
> > I guess nobody cares about macOS enough to actually spend time on this
> 
> It's not really about time but I don't have access to Mac hardware, and it's not possible to test rendering on virtual machine. Could you test latest build from this repository, `sim` branch, with `USE_IMGUI_GL3`, copy `shaders` directory from repository into your maszyna files, with these .ini settings?
> ```
> python.enabled no
> gfxrenderer full
> gfx.skippipeline yes
> gfx.shadergamma yes
> ```

I tried the latest commit from your branch with suggested settings, still the same, starts fine, GUI works, nothing on screen. I could potentially provide a test machine if you think it's worth the effort, but to be honest I'm fine using the legacy renderer. I installed Windows 10 as a secondary system to see if the new renderer makes a big difference and sure lights and shadows are nicer but it comes at a pretty large cost of performance. 

> What do you mean? MaSzyna 20.09 pack contained some older build from this repo, but it is not worth using old executables now. SVN contains newer game assets but they are not much relevant in debugging this.

Right, I just didn't really know how the code version here relates to what is in the public release. I tried your latest binaries on Windows and they worked fine, so it's no longer relevant. Just wanted to rule out the possible change made since release that would make it work on Win but not on macOS.

> Kernel crashes could have security implications, so I guess Apple _should_ care about them.

I have reported some bugs to their dev team in 2016  in person when I went to one their WWDC dev conference (including a minimum reproduction code) now almost 5 years later and nothing happened about it,, so while I understand they should care I know nothing's going to happen. Especially as I'm using a 4 year old computer, they'll just suggest to upgrade :)

> It might have been some extraneous includes in imgui2 builds, I think I fixed this.

I can confirm it builds fine without explicit OpenGL requirement now, thanks!

> I also fixed Azure Pipelines builds and they should contain MacOS builds with any extra dependencies linked statically. While at it I have added ARM64 builds too (for Apple Silicon hardware).

Cool! I found the artifacts here: https://dev.azure.com/milek7/maszyna/_build/results?buildId=376&view=artifacts&pathAsName=false&type=publishedArtifacts but it doesn't look like it's possible to actually download anything. Login required?
 
> Overall, everything this PR contained except mouse scaling should be already addressed by recent changes in this repository.

I found an alternative workaround for that. Adding a GLFW_COCOA_RETINA_FRAMEBUFFER window hint with value set to FALSE disables retina scaling. Of course then you need to specify the resolution in real pixels rather than scaled ones,, so it's different for each display but essentially the effect is the same.

piopawlu committed:
Gated glfwGetWindowContentScale with endif checking glfw version

Milek7 commented:
> Cool! I found the artifacts here: https://dev.azure.com/milek7/maszyna/_build/results?buildId=376&view=artifacts&pathAsName=false&type=publishedArtifacts but it doesn't look like it's possible to actually download anything. Login required?

No login required, there's hidden menu at 3 small dots on the right side :P I just added extra build with imgui gl2, it might be useful for you.

> I found an alternative workaround for that. Adding a GLFW_COCOA_RETINA_FRAMEBUFFER window hint with value set to FALSE disables retina scaling. Of course then you need to specify the resolution in real pixels rather than scaled ones,, so it's different for each display but essentially the effect is the same.

I think setting GLFW_COCOA_RETINA_FRAMEBUFFER to FALSE will scale render resolution, resulting in linear upscaling from lower than native resolution, no? That might be undesirable.

piopawlu commented:
> No login required, there's hidden menu at 3 small dots on the right side :P I just added extra build with imgui gl2, it might be useful for you.

Thanks! I can confirm the build you published (imgui2) works fine in legacy mode. I will be syncing my branch with yours as I have some custom changes to support my hardware controller, but great that there's a nightly build for MacOS now!

> I think setting GLFW_COCOA_RETINA_FRAMEBUFFER to FALSE will scale render resolution, resulting in linear upscaling from lower than native resolution, no? That might be undesirable.

Maybe it would be worth adding a config option to support this. I think in some cases it may be useful to intentionally enable upscaling. Without that I am forced to run the game at 4K resolution if I want full screen (legacy renderer). I think the modern would allow me to customise framebuffer size if I got it to work :)

Milek7 commented:
> Maybe it would be worth adding a config option to support this. I think in some cases it may be useful to intentionally enable upscaling. Without that I am forced to run the game at 4K resolution if I want full screen (legacy renderer). I think the modern would allow me to customise framebuffer size if I got it to work :)

@piopawlu: Modern renderer allows different render resolution in full pipeline mode. But yes, for other modes (legacy renderer or modern with skippipeline) that might be useful.

Modern renderer should be fixed by https://github.com/Milek7/maszyna/commit/78a15bf527a3a0830d83bef0f8dead8095fae2bc

piopawlu commented:
> Modern renderer should be fixed by [78a15bf](https://github.com/Milek7/maszyna/commit/78a15bf527a3a0830d83bef0f8dead8095fae2bc)

🎉 Indeed, working fine! I still don't understand how it worked fine on other platforms, but the most important thing is that it's fixed. Thanks for spending time on this! 🍺

Milek7 commented:
I reverted this because it caused problems on other platforms.
I implemented alternate fix in 90a6b7c3b5cf9d436a238723304c4a832c818c5f, could you check if this works fine on retina?

piopawlu commented:
Yeah, it works fine both in full screen and window mode. UI and train controls can be operated without any issues. I found a different edge case but it's really invalid config so not sure if you want to prevent this. Basically if you specify the frame buffer size larger than the window size and enable full screen mode it gets crazy and end up looking like this:

![Screenshot 2021-01-25 at 09 28 35](https://user-images.githubusercontent.com/624030/105680494-1e648c00-5ef0-11eb-904f-29ce8f63a9b7.png)

Mouse is then not working correctly, it is still using the entire screen area but operating in the tiny window in bottom left corner. I don't think it's anything new, and maybe just exiting with an error if config fb size > window size would be enough.
